### PR TITLE
[devicelab] Cocoon client

### PR DESCRIPTION
**Pull request by CaseyHillers**
## Description

Ports logic from flutter/cocoon agent to be directly included in the test runner script.

If `service-account-file` and `task-key` are passed on devicelab runs, it will attempt to upload the results to Cocoon.

This is needed for migrating the devicelab to LUCI as we are deprecating the Cocoon agent.

## Related Issues

https://github.com/flutter/flutter/issues/66191 - Add support to devicelab test runner to upload results to Cocoon

## Tests

I added the following tests:

Cocoon checks for validating the sent request and reading the service account file correctly.

Runner invokes cocoon upload when a service account file flag is passed.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: dev/devicelab/bin/run.dart [hunk 1]
```diff
@@ -35,6 +36,16 @@ bool exitOnFirstTestFailure;
...
 
+/// File containing a service account token.
+///
+/// If passed, the test run results will be recorded by Flutter infrastructure.
+String serviceAccountFile;
+
+/// Key for the task to upload to in Flutter infrastructure.
+/// 
+/// Only used if [serviceAccountFile] is passed.
+String taskKey;
```
Comment (CaseyHillers) [inline dev/devicelab/bin/run.dart:47]: I think we have all the resources needed to compute the task key here, but then we make it harder to change it on the backend.

**Review by godofredoc — APPROVED** [review]

**Review by jonahwilliams — APPROVED** [review]
Nice! LGTM

Also: https://github.com/flutter/flutter/issues/68240

File: dev/devicelab/bin/run.dart (hunk 1 above)
Comment (godofredoc) [reply dev/devicelab/bin/run.dart:47]: Is it possible to only use the commit?

Comment (CaseyHillers) [reply dev/devicelab/bin/run.dart]: We can pass the commit instead, and update cocoon to take the commit sha + task name.

Comment (CaseyHillers) [reply dev/devicelab/bin/run.dart]: Can we assume git will be on the host running this test? If so, we could just pull this from git

File: dev/devicelab/bin/run.dart [hunk 2]
```diff
@@ -35,6 +36,16 @@ bool exitOnFirstTestFailure;
 /// The device-id to run test on.
 String deviceId;
 
+/// File containing a service account token.
+///
+/// If passed, the test run results will be recorded by Flutter infrastructure.
+String serviceAccountFile;
```
[the remark this answers is not in the document]
Comment (CaseyHillers) [reply dev/devicelab/bin/run.dart]: Done

File: dev/devicelab/lib/framework/cocoon.dart [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import 'task_result.dart';
+import 'utils.dart';
+
+/// Class for test runner to interact with Flutter's infrastructure service, Cocoon.
+///
+/// Cocoon assigns bots to run these devicelab tasks on real devices.
+/// To retrieve these results, the test runner needs to send results back so so the database can be updated.
+class Cocoon {
+  Cocoon({
+    String serviceAccountPath,
```
[the remark this answers is not in the document]
Comment (CaseyHillers) [reply dev/devicelab/lib/framework/cocoon.dart]: Made to serviceAccountTokenPath

File: dev/devicelab/lib/framework/utils.dart [hunk 4]
```diff
@@ -725,3 +725,34 @@ Future<int> gitClone({String path, String repo}) async {
         () => exec('git', <String>['clone', repo]),
   );
 }
+
+/// Call [fn] retrying so long as [retryIf] return `true` for the exception
```
[the remark this answers is not in the document]
Comment (CaseyHillers) [reply dev/devicelab/lib/framework/utils.dart:729]: I had this discussion on discord, and consensus is for something trivial like this to just implement it. It prevents future headaches.

Initially, I did try it and kept running into headaches with getting it to not have pub issues.

File: dev/devicelab/test/cocoon_test.dart [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      final AuthenticatedCocoonClient client = AuthenticatedCocoonClient(serviceAccountPath, filesystem: fs);
+      expect(client.serviceAccountToken, serviceAccountToken);
+    });
+
+    test('throws error when service account file not found', () {
+      final AuthenticatedCocoonClient client = AuthenticatedCocoonClient('idontexist', filesystem: fs);
+      expect(() => client.serviceAccountToken, throwsA(isA<FileSystemException>()));
+    });
+  });
+}
```
[the remark this answers is not in the document]
Comment (CaseyHillers) [reply dev/devicelab/test/cocoon_test.dart]: Done

File: dev/devicelab/bin/run.dart (hunk 1 above)
Comment (CaseyHillers) [reply dev/devicelab/bin/run.dart]: I started to work on the backend change for this, and realized we're going to need to send the branch as well.

Do you know if recipes knows about the branch?

Comment (godofredoc) [reply dev/devicelab/bin/run.dart]: They now about the release ref dev, stable, etc. But I assume getting the branch is possible using git.

Comment (CaseyHillers) [reply dev/devicelab/bin/run.dart]: I got a bit further in the backend, and we can scratch pulling the branch here. I removed the git commit sha flag, and pull it in the cocoon client via git.

If we switch to building these tests as a binary for the bot, we'll need to pass the commit sha as a flag.

**Review by godofredoc — APPROVED** [review]
Thanks for implementing this functionality.

**Pull request merged**